### PR TITLE
feat: Retrieve all job listings 

### PR DESCRIPTION
**Pull request by Asin-Junior-Honore**
### Pull Request: [FEAT] - External Dynamic Pages - Job Listings - Retrieve All Job Listings #145

**Description**

Implemented a feature to retrieve all job listings from the database. This endpoint allows users to fetch job listings with pagination. The response includes job details and pagination information.

**Endpoints**

- `GET /api/v1/jobs`

**Response Format**

```json
{
  "message": "Job listings retrieved successfully.",
  "data": [
    {
      "title": "String",
      "description": "String",
      "location": "String",
      "salary": "String",
      "job_type": "String"
    }
  ],
  "pagination": {
    "current_page": "Int",
    "total_pages": "Int",
    "page_size": "Int",
    "total_items": "Int"
  }
}
```

**Manual Testing**

1. **Retrieve All Job Listings**

   - **Action**: Send a request to the `GET /api/v1/jobs` endpoint.
   - **Expected Result**: You should receive a response with the job listings, including the `message`, `data`, and `pagination` fields.

   **Example Request:**
   ```bash
   curl -X GET http://localhost:3000/api/v1/jobs
   ```

2. **Test Pagination**

   - **Action**: Send a request to the `GET /api/v1/jobs` endpoint with query parameters for pagination (e.g., `page` and `pageSize`).
   - **Expected Result**: The response should include the `pagination` details with `current_page`, `total_pages`, `page_size`, and `total_items`.

   **Example Request:**
   ```bash
   curl -X GET "http://localhost:3000/api/v1/jobs?page=2&pageSize=5"
   ```

**Checklist**

- [x] Implemented an endpoint to retrieve all job listings.
- [x] Added pagination to the job listings endpoint.
- [x] Validated response format includes `message`, `data`, and `pagination`.
- [x] Ensured proper error handling for unexpected issues.

**Reference Issue**

[FEAT] - External Dynamic pages - Job Listings - Retrieve All Job Listings

---

### CODE REVIEW

**Comment by Asin-Junior-Honore** [comment]
@Cobidex 
@Heba-WebDev 
I have implemented the requested changes please review 🙏

File: src/modules/jobs/create-job.dto.ts [hunk 1]
```diff
@@ -0,0 +1,21 @@
+import { IsString } from 'class-validator';
+
+export class CreateJobDto {
+  @IsString()
```
Comment (Cobidex) [inline src/modules/jobs/create-job.dto.ts:4]: Add docs to your dto for swagger

File: src/modules/jobs/job.controller.ts [hunk 2]
```diff
@@ -0,0 +1,12 @@
+import { Controller, Get, Query } from '@nestjs/common';
+import { JobService } from './job.service';
+
+@Controller('api/v1/jobs')
```
Comment (Cobidex) [inline src/modules/jobs/job.controller.ts:4]: There is global path prefix set. Remove api/v1/.  Only jobs is required

Also add docs

File: src/modules/jobs/job.controller.ts [hunk 3]
```diff
@@ -0,0 +1,12 @@
...
+export class JobController {
+  constructor(private readonly jobService: JobService) {}
+
+  @Get()
+  async getAllJobs(@Query('page') page = 1, @Query('pageSize') pageSize = 10): Promise<any> {
```
Comment (Cobidex) [inline src/modules/jobs/job.controller.ts:9]: Add docs

File: src/modules/jobs/job.service.ts [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+import { Job } from './job.entity';
+
+@Injectable()
+export class JobService {
+  constructor(
+    @InjectRepository(Job)
+    private jobRepository: Repository<Job>
+  ) {}
+
+  async findAll(page: number = 1, pageSize: number = 10): Promise<any> {
```
Comment (Cobidex) [inline src/modules/jobs/job.service.ts:13]: Wrap your points of db in try/catch block for proper error logging.

Reference auth services for guidance

Comment (Cobidex) [inline src/modules/jobs/job.service.ts:13]: Your return values and promises should have a shape.  Example: Promise<Job[]>

This helps catch errors before run-time, improving quality enables the full power of typescript

File: src/modules/jobs/job.controller.ts (hunk 3 above)
Comment (Asin-Junior-Honore) [reply src/modules/jobs/job.controller.ts]: Done

File: src/modules/jobs/job.service.ts [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+        data: result,
+        pagination: {
+          current_page: page,
+          total_pages: totalPages,
+          page_size: pageSize,
+          total_items: total,
+        },
+      };
+    } catch (error) {
+      console.error('Error retrieving job listings:', error);
```
Comment (Cobidex) [inline src/modules/jobs/job.service.ts:34]: Use the logger provided by Nestjs for this. Reference Authservice file again for clarification on usage

**Review by Cobidex — APPROVED** [review]

File: src/modules/jobs/job.entity.ts
Comment (Heba-WebDev) [inline src/modules/jobs/job.entity.ts:1]: Specify the nullability of a column please.

File: src/modules/jobs/jobs.controller.spec.ts
Comment (Heba-WebDev) [inline src/modules/jobs/jobs.controller.spec.ts:1]: It's better to extract the mock and have it somewhere else for better maintainability.

**Review by Heba-WebDev — CHANGES_REQUESTED** [review]
Left you few comments

File: src/modules/jobs/job.entity.ts
Comment (Asin-Junior-Honore) [reply src/modules/jobs/job.entity.ts:1]: Done

File: src/modules/jobs/jobs.controller.spec.ts
Comment (Asin-Junior-Honore) [reply src/modules/jobs/jobs.controller.spec.ts:1]: All done and fixed

**Comment by Asin-Junior-Honore** [comment]
@Heba-WebDev please the requested changes has been made 
@buka4rill 
@Homoakin619

**Review by Heba-WebDev — APPROVED** [review]

File: .vscode/settings.json [hunk 6]
```diff
@@ -10,9 +10,14 @@
     "statusBar.background": "#1a1f15",
     "statusBarItem.hoverBackground": "#333d2a",
     "statusBar.foreground": "#e7e7e7",
-    "panel.border": "#333d2a",
-    "sideBar.border": "#333d2a",
-    "editorGroup.border": "#333d2a"
+    "commandCenter.border": "#e7e7e799",
```
Comment (Homoakin619) [inline .vscode/settings.json:13]: revert these changes, don't touch settings

File: src/modules/jobs/job.service.ts [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+@Injectable()
+export class JobService {
+  private readonly logger = new Logger(JobService.name);
+
+  constructor(
+    @InjectRepository(Job)
+    private jobRepository: Repository<Job>
+  ) {}
+
+  async findAll(page: number = 1, pageSize: number = 10): Promise<PaginationResult<Job>> {
```
Comment (Homoakin619) [inline src/modules/jobs/job.service.ts:16]: make this more descriptive like saying getAllJobs

Comment (Asin-Junior-Honore) [reply src/modules/jobs/job.service.ts]: Done sir 🤧

File: .vscode/settings.json (hunk 6 above)
Comment (Asin-Junior-Honore) [reply .vscode/settings.json]: Done

**Review by Cobidex — APPROVED** [review]
LGTM

**Review by davidolanrewaju — APPROVED** [review]
